### DISCUSSION TITLE
Recover password token is invalidated by pre-fetching.

### DISCUSSION THREAD

**Discussion by elvezpablo** [ory/kratos#2530] [Q&A] [answered]
We recently ran into an issue where our recovery password emails worked in dev and stage but not in production. The issue is that the RecoveryUrl GET request is called (in this case by GMail) before the user opens the email [1][2][3]. Since our public Kratos is only available in production it only happens in production. 

The solution we are working on will introduce a page before the RecoveryUrl to avoid the token invalidation and then either redirect via JS or instruct the user to click. Is this the recommended approach to this flow or have others found a better solution to this?

(happy to elaborate if needed)

1. https://news.ycombinator.com/item?id=28240279
3. https://support.google.com/mail/thread/16878288/gmail-is-opening-and-caching-urls-within-emails-without-user-intervention-how-and-why?hl=en
4. https://www.sparkpost.com/blog/gmail-prefetching-images/

**Comment by aeneasr** [comment, marked as answer]
Thank you Paul, that is indeed an issue! I think a better solution to this problem is to use codes instead of magic links - so we send a code to the user they have to input in the UI. I started with an implementation but currently have other prios…

https://github.com/ory/kratos/pull/2184

Please feel free to continue this work!

**Reply by elvezpablo** [reply]
Awesome thanks for the confirmation. We would love to use the codes and appreciate the efforts you all put in.
